tests/test_unified_install_cache.py

```diff
@@ -0,0 +1,135 @@
+  ptex_cache = os.path.join(tmpdir, '.ptex')
+  resolver_settings['cache'] = ptex_cache
+
+  resolved_distributions = resolver.resolve(
+    requirements=[repo_root] + configuration['requirements'],
+    **resolver_settings
+  )
+
+  pex_info = PexInfo.default()
+  pex_info.pex_root = ptex_cache
+  pex_builder = PEXBuilder(pex_info=pex_info)
+
+  dehydrated_requirements = []
+  for resolved_distribution in resolved_distributions:
+    requirement = resolved_distribution.requirement
+    if requirement.key == 'pex':
+      pex_builder.add_requirement(requirement)
+      pex_builder.add_distribution(resolved_distribution.distribution)
+    else:
+      dehydrated_requirements.append(str(requirement))
+  configuration['requirements'] = dehydrated_requirements
+
+  with open(os.path.join(tmpdir, 'PTEX-INFO'), 'w') as fp:
+    json.dump(configuration, fp)
+  pex_builder.add_resource(fp.name, os.path.basename(fp.name))
+
+  with open(os.path.join(tmpdir, 'ptex.py'), 'w') as fp:
+    fp.write(dedent("""
+      import contextlib
+      import json
+      import os
+      import sys
+      import zipfile
+
+      from pex import resolver
+      from pex.pex_builder import PEXBuilder
+      from pex.pex_info import PexInfo
+      from pex.util import CacheHelper
+      from pex.variables import ENV
+
+      self = sys.argv[0]
+      ipex_file = '{}.ipex'.format(os.path.splitext(self)[0])
+
+      if not os.path.isfile(ipex_file):
+        print('Hydrating {} to {}'.format(self, ipex_file))
+
+        with contextlib.closing(zipfile.ZipFile(self)) as zf:
+          ptex_info = json.load(zf.open('PTEX-INFO'))
+        resolver_settings = ptex_info['resolver_settings']
+
+        # Perform a fully pinned intransitive resolve to hydrate the install cache (not the
+        # pex!).
+        resolved_distributions = resolver.resolve(
+          requirements=ptex_info['requirements'],
+          transitive=False,
+          **resolver_settings
+        )
+
+        # Now populate the pex with the pinned requirements and distribution names and hashes
+        # only.
+        pex_info = PexInfo.default()
+        pex_info.pex_root = resolver_settings['cache']
+        pex_builder = PEXBuilder(pex_info=pex_info)
+        for resolved_distribution in resolved_distributions:
+          requirement = resolved_distribution.requirement
+          pex_builder.add_requirement(requirement)
+
+          distribution = resolved_distribution.distribution
+          dist_name = os.path.basename(distribution.location)
+          dist_hash = CacheHelper.dir_hash(distribution.location)
+          pex_info.add_distribution(dist_name, dist_hash)
+        pex_builder.set_entry_point(ptex_info['entry_point'])
+        pex_builder.build(ipex_file)
+
+      os.execv(ipex_file, [ipex_file] + sys.argv[1:])
+    """))
+  pex_builder.set_executable(fp.name)
+
+  ptex_file = os.path.join(tmpdir, 'isort.ptex')
+  pex_builder.build(ptex_file)
```

Comment:
This here is a plain old PEX. It is the deployable and runnable unit just like fat pexes. It so happens this PEX, when run, has user code that uses the PEX API to generate a new (thin) PEX on the host and delegate to that - but that's all opaque to the deployer / runner.